File: ncRepositorios/pCaracteristicaItem.cs
using System;
using System.Collections.Generic;
using Dapper;
using MySql.Data.MySqlClient;
using ncNComum;
using ncModelos;

namespace ncRepositorios
{
    public class pCaracteristicaItem : RepositorioBase, IpCaracteristicaItem
    {
        public ColecaoCaracteristicaItem Listar()
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var lista = conn.Query<dCaracteristicaItem>("SELECT cid, caracteristicas_cid, valor FROM caracteristicaitem").AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoCaracteristicaItem();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Listar CaracteristicaItem [" + ToString() + "] - " + ex.Message);
            }
        }

        public ColecaoCaracteristicaItem Consultar(dCaracteristicaItem dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var conditions = new List<string>();
                    var p = new DynamicParameters();
                    if (dados.cid != 0) { conditions.Add("cid=@cid"); p.Add("cid", dados.cid); }
                    if (dados.caracteristicas_cid != 0) { conditions.Add("caracteristicas_cid=@caracteristicas_cid"); p.Add("caracteristicas_cid", dados.caracteristicas_cid); }
                    if (!string.IsNullOrEmpty(dados.valor)) { conditions.Add("valor=@valor"); p.Add("valor", dados.valor); }
                    var where = conditions.Count > 0 ? "WHERE " + string.Join(" AND ", conditions) : "";
                    var sql = $"SELECT cid, caracteristicas_cid, valor FROM caracteristicaitem {where}";
                    var lista = conn.Query<dCaracteristicaItem>(sql, p).AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoCaracteristicaItem();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Consultar CaracteristicaItem [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Incluir(dCaracteristicaItem dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    conn.Execute("INSERT INTO caracteristicaitem (caracteristicas_cid, valor) VALUES (@caracteristicas_cid, @valor)",
                        new { dados.caracteristicas_cid, dados.valor });
                    return (int)conn.ExecuteScalar<long>("SELECT LAST_INSERT_ID()");
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Incluir CaracteristicaItem [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Alterar(dCaracteristicaItem dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    return conn.Execute("UPDATE caracteristicaitem SET caracteristicas_cid=@caracteristicas_cid, valor=@valor WHERE cid=@cid",
                        new { dados.caracteristicas_cid, dados.valor, dados.cid });
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Alterar CaracteristicaItem [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Excluir(dCaracteristicaItem dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    return conn.Execute("DELETE FROM caracteristicaitem WHERE cid=@cid", new { dados.cid });
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Excluir CaracteristicaItem [" + ToString() + "] - " + ex.Message);
            }
        }
    }
}

File: ncPersistencia/pCaracteristicaItem.vb
Imports ncDados.nsCaracteristica
Imports ncComum.nsAcessoBD
Imports ncComum.nsFuncoes
Imports ncComum.nsExcecao

Namespace nsCaracteristica

  Public Class pCaracteristicaItem

    Public Function Listar() As ColecaoCaracteristicaItem

      Dim retorno As ColecaoCaracteristicaItem
      Dim acessoBanco As cAcessoBD
      Dim ds As DataSet
      Dim dt As DataTable
      Dim row As DataRow
      Dim item As dCaracteristicaItem
      Dim comandoSQL As String

      Try

        acessoBanco = New cAcessoBD


        comandoSQL = " Select cid, caracteristicas_cid, valor From caracteristicaitem "

        ds = acessoBanco.ExecutarDS(comandoSQL)

        If Not ds Is Nothing Then
          If ds.Tables.Count > 0 Then
            dt = ds.Tables(0)

            If dt.Rows.Count > 0 Then
              retorno = New ColecaoCaracteristicaItem

              For Each row In dt.Rows
                item = New dCaracteristicaItem

                item.cid = cFuncoes.RetornarInteiro(row("cid"))
                item.caracteristicas_cid = cFuncoes.RetornarInteiro(row("caracteristicas_cid"))
                item.valor = cFuncoes.RetornarTexto(row("valor"))

                retorno.Add(item)
              Next
            Else
              retorno = Nothing
            End If
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Listar CaracteristicaItem [" & Me.ToString() & "] - " & ex.Message, ex)

      End Try

      Listar = retorno

    End Function

    Public Function Consultar(ByVal dados As dCaracteristicaItem) As ColecaoCaracteristicaItem

      Dim retorno As ColecaoCaracteristicaItem
      Dim acessoBanco As cAcessoBD
      Dim ds As DataSet
      Dim dt As DataTable
      Dim row As DataRow
      Dim item As dCaracteristicaItem
      Dim sqlSelect As String
      Dim sqlWhere As String
      Dim sqlFrom As String

      Try

        acessoBanco = New cAcessoBD


        sqlSelect = " Select cid, caracteristicas_cid, valor "
        sqlWhere = String.Empty
        sqlFrom = " From caracteristicaitem "

        '-- cid
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.cid, "cid")

        '-- caracteristicas_cid
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.caracteristicas_cid, "caracteristicas_cid")

        '-- valor
        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.valor, "valor")

        If Not sqlWhere.Equals(String.Empty) Then
          sqlWhere = " WHERE " & sqlWhere
        End If

        ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere)

        If Not ds Is Nothing Then
          If ds.Tables.Count > 0 Then
            dt = ds.Tables(0)

            If dt.Rows.Count > 0 Then
              retorno = New ColecaoCaracteristicaItem

              For Each row In dt.Rows
                item = New dCaracteristicaItem

                item.cid = cFuncoes.RetornarInteiro(row("cid"))
                item.caracteristicas_cid = cFuncoes.RetornarInteiro(row("caracteristicas_cid"))
                item.valor = cFuncoes.RetornarTexto(row("valor"))

                retorno.Add(item)
              Next
            Else
              retorno = Nothing
            End If
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Consultar CaracteristicaItem [" & Me.ToString() & "] - " & ex.Message, ex)

      End Try

      Consultar = retorno

    End Function

    Public Function Incluir(ByVal dados As dCaracteristicaItem) As Integer

      Dim retorno As Integer
      Dim acessoBanco As cAcessoBD
      Dim comandoSQL As String

      Try

        acessoBanco = New cAcessoBD


        comandoSQL = " INSERT INTO " & _
            " caracteristicaitem ( caracteristicas_cid, valor ) " & _
            " VALUES (" & _
            cFuncoes.PersistirTexto(dados.caracteristicas_cid) & "," & _
            cFuncoes.PersistirTexto(dados.valor) & ")"

        retorno = acessoBanco.ExecutarINT(comandoSQL)

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Incluir CaracteristicaItem [" & Me.ToString() & "] - " & ex.Message, ex)

      End Try

      Incluir = retorno

    End Function

    Public Function Alterar(ByVal dados As dCaracteristicaItem) As Integer

      Dim retorno As Integer
      Dim acessoBanco As cAcessoBD
      Dim comandoSQL As String

      Try

        acessoBanco = New cAcessoBD


        comandoSQL = " UPDATE caracteristicaitem SET " & _
            " caracteristicas_cid = " & cFuncoes.PersistirTexto(dados.caracteristicas_cid) & "," & _
            " valor = " & cFuncoes.PersistirTexto(dados.valor) & _
            " WHERE " & _
            " cid = " & dados.cid.ToString()

        retorno = acessoBanco.ExecutarINT(comandoSQL)

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Alterar CaracteristicaItem [" & Me.ToString() & "] - " & ex.Message, ex)

      End Try

      Alterar = retorno

    End Function

    Public Function Excluir(ByVal dados As dCaracteristicaItem) As Integer

      Dim retorno As Integer
      Dim acessoBanco As cAcessoBD
      Dim comandoSQL As String

      Try

        acessoBanco = New cAcessoBD

        comandoSQL = " DELETE FROM caracteristicaitem " & _
            " WHERE cid = " & dados.cid.ToString()

        retorno = acessoBanco.ExecutarINT(comandoSQL)

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Excluir CaracteristicaItem [" & Me.ToString() & "] - " & ex.Message, ex)

      End Try

      Excluir = retorno

    End Function

  End Class

End Namespace
